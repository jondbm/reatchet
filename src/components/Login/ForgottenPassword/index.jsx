import React, { Component } from 'react'
import styled from 'styled-components'

import TextInput from '../../common/TextInput'
import Button from '../../common/Buttons/Button'
import { Wrapper, Title, Subtitle, StyledRecaptcha } from '../sharedComponents'

import { validateEmail } from '../../../helpers/validation'
import theme from '../../../theme'

/*
*****************
STYLED COMPONENTS
*****************
*/

const Footer = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  align-self: stretch;
  margin-top: ${theme.spacing.sm};
`

const SubmitButton = styled(Button)`
  margin-left: ${theme.spacing.xs};
`

/*
*********
COMPONENT
*********
*/

class ForgottenPassword extends Component {
  state = {
    email: '',
    emailIsValid: true,
    recaptchaToken: ''
  }

  updateEmail = value => {
    const isValid = validateEmail(value)

    this.setState({
      email: value,
      emailIsValid: isValid
    })
  }

  handleRecaptcha = token => {
    this.setState({ recaptchaToken: token })
  }

  handleExpiredRecaptcha = () => {
    this.setState({ recaptchaToken: '' })
  }

  // can't seem to spy on history.goBack()
  /* istanbul ignore next */
  handleCancel = () => {
    this.props.history.goBack()
  }

  render() {
    const { email, emailIsValid } = this.state
    return (
      <Wrapper>
        <Title>Forgotten your password?</Title>
        <Subtitle>
          Pop your email address in below and in return weâ€™ll send you a
          password reset link!
        </Subtitle>
        <TextInput
          label="Email address"
          handleChange={this.updateEmail}
          value={email}
          isValid={emailIsValid}
          validationMessage="Enter a valid email address"
        />
        <StyledRecaptcha
          handleRecaptcha={this.handleRecaptcha}
          handleExpiredRecaptcha={this.handleExpiredRecaptcha}
          id="forgotten-password"
        />
        <Footer>
          <Button
            data-el="cancel-button"
            variant="secondary"
            onClick={this.handleCancel}
          >
            Cancel
          </Button>
          <SubmitButton data-el="submit-button">Submit</SubmitButton>
        </Footer>
      </Wrapper>
    )
  }
}

export default ForgottenPassword
