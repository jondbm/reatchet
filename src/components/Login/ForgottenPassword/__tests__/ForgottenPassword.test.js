import React from 'react'
import 'jest-styled-components'
import createMemoryHistory from 'history/createMemoryHistory'
import { shallow } from 'enzyme'

import ForgottenPassword from '../'

describe('<ForgottenPassword /> component', () => {
  let component

  const memoryHistory = createMemoryHistory({
    initialEntries: ['/login', '/login/forgotten-password'],
    initialIndex: 0,
    keyLength: 2,
    getUserConfirmation: null
  })

  beforeEach(() => {
    component = shallow(<ForgottenPassword history={memoryHistory} />)
  })

  describe('basic rendering', () => {
    it('should correctly render the component', () => {
      expect(component.exists()).toBe(true)
    })

    it('should match the snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })

  describe('updateEmail()', () => {
    let updateEmail
    beforeEach(() => {
      updateEmail = component.instance().updateEmail
    })

    it('should set the correct email address in state', () => {
      updateEmail('test@test.com')
      component.update()
      expect(component.state('email')).toBe('test@test.com')
    })

    it('should set state emailIsValid when it is valid', () => {
      updateEmail('test@test.com')
      component.update()
      expect(component.state('emailIsValid')).toBe(true)
    })

    it('should set state !emailIsValid when it is invalid', () => {
      updateEmail('test@test')
      component.update()
      expect(component.state('emailIsValid')).toBe(false)
    })
  })

  describe('handleRecaptcha()', () => {
    it('should set the recaptcha token in state', () => {
      const handleRecaptcha = component.instance().handleRecaptcha
      handleRecaptcha('token123')
      component.update()
      expect(component.state('recaptchaToken')).toBe('token123')
    })
  })

  describe('handleExpiredRecaptcha()', () => {
    it('should remove the recaptcha token from state', () => {
      const handleExpiredRecaptcha = component.instance().handleExpiredRecaptcha
      handleExpiredRecaptcha()
      component.update()
      expect(component.state('recaptchaToken')).toBe('')
    })
  })
})
