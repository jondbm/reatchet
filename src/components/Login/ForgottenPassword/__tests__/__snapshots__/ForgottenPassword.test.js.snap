// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgottenPassword /> component basic rendering should match the snapshot 1`] = `
<sharedComponents__Wrapper>
  <sharedComponents__Title>
    Forgotten your password?
  </sharedComponents__Title>
  <sharedComponents__Subtitle>
    Pop your email address in below and in return weâ€™ll send you a password reset link!
  </sharedComponents__Subtitle>
  <TextInput
    handleChange={[Function]}
    isPasswordInput={false}
    isValid={true}
    label="Email address"
    validationMessage="Enter a valid email address"
    value=""
  />
  <sharedComponents__StyledRecaptcha
    handleExpiredRecaptcha={[Function]}
    handleRecaptcha={[Function]}
    id="forgotten-password"
  />
  <ForgottenPassword__Footer>
    <Button
      data-el="cancel-button"
      data-key=""
      inverse={false}
      onClick={[Function]}
      variant="secondary"
    >
      Cancel
    </Button>
    <ForgottenPassword__SubmitButton
      data-el="submit-button"
    >
      Submit
    </ForgottenPassword__SubmitButton>
  </ForgottenPassword__Footer>
</sharedComponents__Wrapper>
`;
